switchbot/__init__.py

```diff
@@ -93,119 +149,156 @@ def _writekey(self, key) -> bool:
-            "Cannot connect to Switchbot. Retrying (remaining: %d)...", retry
-        )
+        _LOGGER.warning("Cannot connect to Switchbot. Retrying (remaining: %d)", retry)
         time.sleep(DEFAULT_RETRY_TIMEOUT)
         return self._sendcommand(key, retry - 1)
 
-    def get_servicedata(self, retry=DEFAULT_RETRY_COUNT, scan_timeout=5) -> bytearray:
-        """Get BTLE 16b Service Data,
-        returns after the given timeout period in seconds."""
+    def discover(self, retry=DEFAULT_RETRY_COUNT, scan_timeout=5) -> dict:
+        """Find switchbot devices and their advertisement data."""
+
         devices = None
 
-        waiting_time = self._time_between_update_command - time.time()
-        if waiting_time > 0:
-            time.sleep(waiting_time)
         try:
             devices = bluepy.btle.Scanner().scan(scan_timeout)
 
         except bluepy.btle.BTLEManagementError:
-            _LOGGER.warning("Error updating Switchbot.", exc_info=True)
+            _LOGGER.error("Error scanning for switchbot devices", exc_info=True)
 
         if devices is None:
             if retry < 1:
                 _LOGGER.error(
-                    "Switchbot update failed. Stopping trying.", exc_info=True
+                    "Scanning for Switchbot devices failed. Stop trying", exc_info=True
                 )
                 return None
 
             _LOGGER.warning(
-                "Cannot update Switchbot. Retrying (remaining: %d)...", retry
+                "Error scanning for Switchbot devices. Retrying (remaining: %d)",
+                retry,
             )
             time.sleep(DEFAULT_RETRY_TIMEOUT)
-            return self.get_servicedata(retry - 1, scan_timeout)
-
-        for device in devices:
-            if self._mac.lower() == device.addr.lower():
-                for (adtype, _, value) in device.getScanData():
+            return self.discover(retry - 1, scan_timeout)
+
+        for dev in devices:
+            if dev.getValueText(7) == UUID:
+                dev_id = dev.addr.replace(":", "")
+                self._all_services_data[dev_id] = {}
+                self._all_services_data[dev_id]["mac_address"] = dev.addr
+                for (adtype, desc, value) in dev.getScanData():
                     if adtype == 22:
-                        service_data = value[4:].encode()
-                        service_data = binascii.unhexlify(service_data)
+                        _model = binascii.unhexlify(value[4:6]).decode()
+                        if _model == "H":
+                            self._all_services_data[dev_id]["data"] = _process_wohand(
+                                value[4:]
+                            )
+                            self._all_services_data[dev_id]["data"]["rssi"] = dev.rssi
+                            self._all_services_data[dev_id]["model"] = _model
+                            self._all_services_data[dev_id]["modelName"] = "WoHand"
+                        if _model == "c":
```

Comment:
```suggestion
                        elif _model == "c":
```

Reply:
if..elif..else doesn't make sense for these as there might be a new model that we don't account for. The service data might then be invalid for the last else.